### PR TITLE
Zy/329 visual not group more than 20 levels

### LINKED ISSUE #329: VISUAL: Do not allow GROUP BY with more than 20 levels
For the GROUP BY do not list any categoric variables with more than 10 categories.

Image

### PR DESCRIPTION
**Pull request by zheyxu**
# Pull Request Details

## What issue does this PR address

- visual not group more than 20 levels

- Link to associated issue: #329

## Checklist

Complete the check-list below to ensure your branch is ready for PR.

Flutter Style Guide: https://survivor.togaware.com/gnulinux/flutter-style.html

- [x] Screenshots included in linked issue
- [x] Changes adhere to the style and coding guideline
- [x] No confidential information
- [x] No duplicated content
- [x] No lint check errors related to your changes (`make prep` or `flutter analyze lib`)
- [x] Pre-exisiting lint errors noted: [HERE]
- [x] No integrated test errors
- [x] Tested on at least one device
  - [ ] Android Phone
  - [ ] Android Emulator
  - [ ] Chrome on Android
  - [ ] Chrome
  - [ ] iOS
  - [ ] Linux
  - [x] MacOS
  - [ ] Windows
- [x] Added two reviewers

## Finalising

Once PR discussion is complete and reviewers have approved:

- [ ] Merge dev into the branch
- [ ] Resolve any conflicts
- [ ] Add one line summary into CHANGELOG.md
- [ ] Bump appropriate version number in pubspec.yaml
- [ ] Push to git repository and review
- [ ] Merge PR into dev (gjwgit)

### CODE REVIEW

File: pubspec.yaml
Comment (gjwgit) [inline pubspec.yaml:1]: Why is the dependency override back in? Do you have dart-code-metrics working?

File: lib/utils/get_catergoric.dart [hunk 1]
```diff
@@ -25,19 +25,29 @@
...
   Map<String, Type> roles = ref.read(typesProvider);
 
+  String stdout = ref.watch(stdoutProvider);
+
+  List<String> largeFactors = extractLargeFactors(stdout);
+
   List<String> rtn = [];
   roles.forEach((key, value) {
-    if (value == Type.categoric) {
+    if (value == Type.categoric &&
```
Comment (gjwgit) [inline lib/utils/get_catergoric.dart]: Could you explain the logic here. If I have a variable (like state) that is categoric and I choose to IGNORE it but it does not have largeFactors, then it gets listed for Group By? I think if I chose IGNORE I would not expect it to be listed as GROUP BY?

File: pubspec.yaml
Comment (zheyxu) [reply pubspec.yaml:1]: If dependency not back in, the project fails to configure. I have to make changes to dart-code-metrics.

File: lib/utils/get_catergoric.dart (hunk 1 above)
Comment (zheyxu) [reply lib/utils/get_catergoric.dart]: Hi @gjwgit, code is updated.

File: pubspec.yaml
Comment (gjwgit) [reply pubspec.yaml:1]: Okay/ Let's keep it there for now.

File: lib/utils/get_catergoric.dart [hunk 2]
```diff
@@ -25,22 +25,35 @@
...
 List<String> getCategoric(WidgetRef ref) {
-  // The typesProvider lists the types for the different variables which we
-  // need to know for parsing the R scripts.
+  // The typesProvider lists the types for the different variables which we need
+  // to know for parsing the R scripts.
 
   Map<String, Type> roles = ref.read(typesProvider);
 
-  List<String> rtn = [];
+  String stdout = ref.watch(stdoutProvider);
```
Comment (gjwgit) [inline lib/utils/get_catergoric.dart]: Please add a comment - why watching stdout. Thanks.

**Review by gjwgit — CHANGES_REQUESTED** [review]
Please include a screenshot of the output of `make qtest.tmp` and any follow up standalone tests. Thanks.

File: lib/utils/get_catergoric.dart [hunk 3]
```diff
@@ -41,6 +41,9 @@ List<String> getCategoric(WidgetRef ref) {
 
   Map<String, Type> roles = ref.read(typesProvider);
 
+
+  // Watching stdout to get variables that are Ignored.
```
Comment (gjwgit) [inline lib/utils/get_catergoric.dart]: Why do you need to watch rather than read?

**Comment by zheyxu** [comment]
`integration_test/app_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/dataset_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/dataset_large_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/dataset_reset_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 `Yes ***** First time NO but make dataset_reset.qtest passes`
integration_test/dataset_reset_visual_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/dataset_sherlock_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_correlation_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_correlation_large_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_large_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_missing_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_missing_large_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_set_ignore_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_visual_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/explore_visual_large_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/model_cluster_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/model_nnet_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/model_nnet_large_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/model_tree_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

YES
integration_test/model_tree_large_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

YES
integration_test/transform_cleanup_ignore_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/transform_impute_demo_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES
integration_test/transform_impute_large_test.dart--- xcodebuild: WARNING: Using the first of multiple matching destinations:

 YES`

File: lib/utils/get_catergoric.dart (hunk 3 above)
Comment (zheyxu) [reply lib/utils/get_catergoric.dart]: Thanks. I know `Use ref.watch to make your widgets react to changes in the state, and ref.read when you need to perform actions or access state without triggering rebuilds.` In this case, change this part to `ref.read`.

**Comment by gjwgit** [comment]
> xcodebuild: WARNING: Using the first of multiple matching destinations:

- [ ] It would be nice to get rid of the warning!

**Comment by zheyxu** [comment]
> > xcodebuild: WARNING: Using the first of multiple matching destinations:
> 
> * [ ]  It would be nice to get rid of the warning!

Can be done in other issue.

**Review by gjwgit — APPROVED** [review]

**Pull request merged**